addons/binding/org.openhab.binding.hpprinter/src/main/java/org/openhab/binding/hpprinter/internal/discovery/HPPrinterDiscoveryParticipant.java

```diff
@@ -0,0 +1,158 @@
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.hpprinter.internal.HPPrinterBindingConstants;
+import org.openhab.binding.hpprinter.internal.config.HPPrinterConfiguration;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Discovers printers announced by mDNS. Code modified from ipp binding by Tobias Bräutigam
+ *
+ * @author Stewart Cossey - Initial contribution
+ */
+@NonNullByDefault
+@Component(immediate = true)
+public class HPPrinterDiscoveryParticipant implements MDNSDiscoveryParticipant {
+
+    private final Logger logger = LoggerFactory.getLogger(HPPrinterDiscoveryParticipant.class);
+
+    @Override
+    public Set<ThingTypeUID> getSupportedThingTypeUIDs() {
+        return HPPrinterBindingConstants.SUPPORTED_DEVICES;
+    }
+
+    @Override
+    public String getServiceType() {
+        return "_printer._tcp.local.";
+    }
+
+    @Nullable
+    @Override
+    public ThingUID getThingUID(ServiceInfo service) {
+        if (service.getType() != null) {
+            if (service.getType().equals(getServiceType())) {
+                String uidName = getUIDName(service);
+                ThingTypeUID mdl = findThingType(service);
+
+                return new ThingUID(mdl, uidName);
+            }
+        }
+        return null;
+    }
+
+    private ThingTypeUID findThingType(ServiceInfo service) {
+        String modl = service.getPropertyString("ty").substring(2).replace(" ", "");
+
+        // Lookup Binding type by Binding ID
+        for (ThingTypeUID thingType : HPPrinterBindingConstants.SUPPORTED_DEVICES) {
+            if (thingType.getId().equalsIgnoreCase(modl)) {
+                return thingType;
+            }
+        }
+
+        // If that fails lookup Binding type in the UPNP device lookup table.
+        return HPPrinterBindingConstants.upnpDeviceLookup(modl);
+    }
+
+    private String getUIDName(ServiceInfo service) {
+        return service.getName().replaceAll("[^A-Za-z0-9_]", "_");
+    }
+
+    @Nullable
+    private InetAddress getIpAddress(ServiceInfo service) {
+        InetAddress address = null;
+        for (InetAddress addr : service.getInet4Addresses()) {
+            return addr;
+        }
+        // Fallback for Inet6addresses
+        for (InetAddress addr : service.getInet6Addresses()) {
+            return addr;
+        }
+        return address;
+    }
+
+    @Nullable
+    @Override
+    public DiscoveryResult createResult(ServiceInfo service) {
+        DiscoveryResult result = null;
+        // if service has no data then return nothing
+        if (!service.hasData()) {
+            logger.trace("Service has no data.");
+            return null;
+        }
+        // We only care about HP Printers
+        if (service.getPropertyString("ty").toLowerCase().startsWith("hp")) {
```

Comment:
Would it make sense to store `service.getPropertyString("ty")` in a variable before this line, because the debug (line 115 & 150)/trace (line 153) statement will call the method again (independent of logging level)